third_party/move/move-vm/types/src/utils/union_find.rs

```diff
@@ -0,0 +1,293 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+#[derive(Debug)]
+pub struct UnionFind {
+    parent: Vec<u32>,
+    rank: Vec<u8>,
+    num_sets: usize,
+    capacity: usize,
+}
+
+impl UnionFind {
+    /// Creates a new UnionFind instance with `num_elements` and `capacity`
+    pub fn new_with_size_and_capacity(
+        num_elements: usize,
+        capacity: usize,
+    ) -> Result<Self, &'static str> {
+        if num_elements > std::u32::MAX as usize {
+            return Err("num_elements larger than std::u32::MAX");
+        }
+        if capacity > std::u32::MAX as usize {
+            return Err("capacity larger than std::u32::MAX");
+        }
+        if num_elements > capacity {
+            return Err("num_elements larger than given capacity");
+        }
+
+        let parent = (0..num_elements as u32).collect::<Vec<u32>>();
+        let rank = vec![0u8; num_elements];
+        Ok(Self {
+            parent,
+            rank,
+            num_sets: num_elements,
+            capacity,
+        })
+    }
+
+    /// Creates a new UnionFind instance with given `capacity` and 0 elements
+    pub fn new_with_capacity(capacity: usize) -> Result<Self, &'static str> {
+        Self::new_with_size_and_capacity(0, capacity)
+    }
+
+    /// Creates a new UnionFind instance with `num_elements` and maximal capacity.
+    pub fn new_with_size(num_elements: usize) -> Result<Self, &'static str> {
+        Self::new_with_size_and_capacity(num_elements, std::u32::MAX as usize)
+    }
+
+    /// Creates a new UnionFind instance with `0` elements and maximal capacity.
+    pub fn new() -> Result<Self, &'static str> {
+        Self::new_with_size_and_capacity(0, std::u32::MAX as usize)
+    }
+
+    // Finds the representative of the set containing `id` without bound checking
+    fn find_set_internal(&mut self, id: u32) -> u32 {
+        let mut i = id;
+        let mut n = id;
+
+        while n != self.parent[n as usize] {
+            n = self.parent[n as usize];
+        }
+
+        while n != i {
```

Comment:
so we are directly shortcutting everything to root?
nits: 
I'd add a comment on what's going on.
i'd write i != n since i is the loop variable here.
also maybe I'd do above:
```
let root = {
    let mut j = id;
    loop to get to the root;
    j
}

and then while i != root..
```